### ISSUE TITLE
[Fuzzer] Race condition at HashJoinFinalizeEvent

### ISSUE THREAD

**Issue by PedroTadim** [duckdb/duckdb#4694]
### What happens?

Run the statements:
```
CREATE SEQUENCE t4;
PRAGMA DEBUG_FORCE_EXTERNAL=1;
SELECT 1 FROM ((SELECT 2) INTERSECT (SELECT 2)) t2(c3) WHERE "currval"('t4') = EXISTS (SELECT 2);
```
It may trigger a heap-use-after-free with the address sanitizer at  duckdb::HashJoinFinalizeEvent::HashJoinFinalizeEvent

### To Reproduce

Run the statements above,

### OS:

Linux

### DuckDB Version:

latest from sources

### DuckDB Client:

Shell

### Full Name:

Pedro Ferreira

### Affiliation:

Huawei

### Have you tried this on the latest `master` branch?

- [X] I agree

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] I agree

**Comment by lnkuiper** [comment]
Nice find! I recently reworked the hash join a bit, so this seems to be happening in a new task that I've spawned.

The race condition is between the main thread calling `ClientContext::EndQueryInternal`, and another thread that calls `event->FinishTask()` at `physical_hash_join.cpp:322`. The object of the race condition is the `Executor` that is responsible for finishing the query. The main thread is ending the query while another thread is still working on it.

The query does not finish nicely, as it yields this error:
```
Error: Serialization Error: currval: sequence is not yet defined in this session
```

So I think the issue here is not with the hash join, but rather with the main thread ending the query early due to the error (and deleting the `Executor` while doing so), while other threads may still be working on tasks, which depend on the `Executor`.

**Comment by PedroTadim** [comment]
Thanks for the explanation! So the solution would be for the main thread to wait for other threads to finish their current tasks and then cancel the whole plan?

**Comment by lnkuiper** [comment]
Yes. Either that, or the tasks need to check whether the plan has been canceled before doing anything else with the `Executor`.

**Comment by PedroTadim** [comment]
Now I have a similar issue with this:

```
CREATE VIEW t1(c1) AS ((SELECT 1) UNION DISTINCT (SELECT 1));
CREATE OR REPLACE TABLE t1 AS ((SELECT 1) UNION ALL ((SELECT 2) INTERSECT ALL (SELECT 2)));
```
src/parallel/pipeline_executor.cpp:10:40: runtime error: member call on address 0x6080000092a0 which does not point to an object of type 'duckdb::PhysicalOperator'.

Also, a view is being replaced by a table. Is this allowed?

**Comment by Mytherin** [comment]
Thanks, I have pushed a fix for that in #4733 

> Also, a view is being replaced by a table. Is this allowed?

No, that is not allowed, which is actually where the bug comes from. The correct handling of the exception was missing.
